src/CoreWCF.NetFramingBase/src/CoreWCF/Channels/ServerFramingDuplexSessionChannel.cs

```diff
@@ -332,8 +332,15 @@ protected override async Task OnSendCoreAsync(Message message, CancellationToken
             allowOutputBatching = message.Properties.AllowOutputBatching;
             Connection.Logger.SendMessage(message);
             messageData = EncodeMessage(message);
-            await Connection.Output.WriteAsync(messageData, token);
-            await Connection.Output.FlushAsync();
+            try
+            {
+                await Connection.Output.WriteAsync(messageData, token);
+                await Connection.Output.FlushAsync();
+            }
+            finally
+            {
+                BufferManager.ReturnBuffer(messageData.Array);
```

Comment:
Looking at the WCF code, the async code semantically does it in the same place (code is all over the place with APM, but it's this class that does the return and not the connection.Write call). This is exactly how I would have fixed this.